Cvalley working notify

CollinValley opened the pull request:
Two commits to change our codebase from using foreverDrain to the more proper task::current().notify(). See the commit messages for more verbosity

CollinValley committed:
Immediate Notify: Use task::current().notify()

In our AsyncElement, we previously used a ForeverDrain element that
wouldn't respond to Async::NotReady, and therefore would continue to
poll the functions non-stop. Now we switch out this functionality by,
when the poll() is about to return an Async::NotReady, rescheduling the
task with Tokio by calling task::current().notify() inside the poll
function.

An aside about Tasks, a task is a unit of abstration over a bunch of
related futures. Tokio has this concept of scheduling tasks, done
through the notify function, which tells the Runtime to try calling
the topmost poll function of whatever Task has had it's notify
function called recently. Think of it like reschedule.

In this case, we call notify before we even return the NotReady,
which is an improvement from the previous use of the forever
drain. Since we yield our computation momentarily, it is still less
than ideal, as we are not driving the notify function using some
notion of when there is new work to complete. This will be
fixed in later efforts.

Some additional reading I found helpful on this matter:
https://docs.rs/futures/0.1.28/futures/task/struct.Task.html
https://cafbit.com/post/tokio_internals/

CollinValley committed:
Remove ForeverDrain: No longer needed

Since we are now using the task::current().notify(), we don't need
to use the foreverDrain hack to keep Tokio from sleeping our
tasks forever. Remove it from the codebase since we can now
just use the exhaustiveDrain.

ddowl commented on src/lib.rs:
```diff
@@ -120,10 +120,10 @@ mod tests {
+        let elem3_drain = ExhaustiveDrain::new(1, Box::new(elem3_link.frontend));
 
-        let elem3_consumer = ForeverDrain::new(1, Box::new(elem3_link.backend));
+        let elem3_consumer = ExhaustiveDrain::new(2, Box::new(elem3_link.backend));
```
Good catch on the drain ID.

ddowl commented on src/lib.rs:
```diff
@@ -82,7 +82,7 @@ mod tests {
 
     #[test]
-    fn two_async_elements_no_waiting() {
+    fn two_async_elements_immediate_notify() {
```
I see the point of changing the test name to signify the change in polling behavior, but that isn't what the `no_waiting` was originally specifying. `no_waiting` was referring to the packet generator input immediately yielding values (i.e. never yielding `Async::NotReady`) until it was empty (then sends `Async::Ready(None)`). To verify that the new `task::notify` behavior works correctly, let's construct a new test called `<element_arrangement>_source_packets_generated_on_interval` that uses a `LinearPacketGenerator` rather than `iter_ok`. 

It would be even nicer if we could specify `context`s like in RSpec so we could say

```
context "packets arrive every #{source_interval} seconds" do
  let packet_generator = LinearIntervalGenerator::new(time::Duration::from_millis(source_interval), 10);

  // ... all test examples that use the packet generator
end
```

I'm not sure this is possible with Rust's test framework though. If contexts don't exist, we'll just have to use long test names, which are only slightly more inconvenient.

CollinValley replied:
Okay, I'll put in another commit to add the Linear Generator test, fixup the naming convention. Standby

CollinValley replied:
I had to push this as a new commit, since we remove the ForeverDrain in the 2nd commit and I did not want to fix the merge conflicts of squashing 1 and 3 together.  I'm not sure about making the tests have a bit more Rspec way, maybe someone else could look into that. I'm sure that the very least we could define common functions.

ddowl commented on src/lib.rs:
```diff
@@ -106,7 +106,7 @@ mod tests {
 
     #[test]
-    fn series_sync_and_async_no_waiting() {
+    fn series_sync_and_async_immediate_notify() {
```
Same as above

CollinValley replied:
Resolved in new, 3rd commit in the stack.

ddowl reviewed (changes requested):
The change is pretty straightforward, but left a few comments on the tests.

I think it's great to have little asides in the commit messages explaining our thought process as we go about building. Someone trying to understand the code will probably reading through commit messages, and having helpful explanations of _why_ we made certain design decisions. I'd even take it a bit further and add in a link to any helpful documentation, if it exists.

There's also a run-on-sentence in the third paragraph.

CollinValley committed:
Improve test coverage and verbosity

This commit adds 3 additional tests, driven by a LinearIntervalGenerator
and clarifies a naming convention.  We also add a short bit of
documentation to the first test to explain the name.  All tests pass at
this commit point.

CollinValley commented:
Fixed the run-on sentence in the 1st commit, and dropped in some links that would be helpful to anyone auditing the code.
